File: pset6/sentimental-credit/credit.py
from cs50 import get_string
import re

# Ask for a number from the user and loop until something greater than 0 is given.
while True:
    number = get_string("Number: ")
    if int(number) > 0:
        break

sum = 0

# Adds the last digit of the credit card to sum.
working = int(number)
while (working > 0):
    ld = working % 10
    sum += ld
    working //= 100


# Calculates the sum of the 2nd last digits multipled by 2 and add to sum (simplifies 2 digit results before adding).
working = int(number) // 10
while (working > 0):
    ld = working % 10
    mbt = ld * 2
    sum += (mbt % 10) + (mbt // 10)
    working //= 100


# Regex to see which card it belongs to.
if (sum % 10) == 0:
    if (re.search("^4([0-9]{15}|[0-9]{12})", number)):
        print("VISA")
    elif (re.search("(^34|^37)[0-9]{13}", number)):
        print("AMEX")
    elif (re.search("^[5][1-5][0-9]{14}", number)):
        print("MASTERCARD")
    else:
        print("INVALID")
else:
    print("INVALID")

File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long creditCard;

    //Asks user for card number and will keep asking until it is given a value greater than 0
    do
    {
        printf("Enter a card number: ");
        scanf("%ld", &creditCard);
    }
    while (creditCard <= 0);

    int sum = 0;
    int counter = 0;
    long divide = 10;

    //Used a working copy of the credit card number to make changes and run calculations
    long workingCopy = creditCard;


    //Adds the last digit to a sum and then divides the number by 100 to get the next one to add. Stops when the number is 0.
    while (workingCopy > 0)
    {
        int lastDigit = workingCopy % 10;
        sum += lastDigit;
        workingCopy /= 100;
    }


    //Takes the credit card number and removes the last digit.
    //Then doubles the digit and adds it together if it is a two digit number.
    //Then adds to the sums and divides by 100 to get the next digit.
    workingCopy = creditCard / 10;
    while (workingCopy > 0)
    {
        int lastDigit = workingCopy % 10;
        int multipleByTwo = lastDigit * 2;
        sum += ((multipleByTwo % 10) + (multipleByTwo / 10));
        workingCopy /= 100;
    }

    //Checks the credit cards length and stores it in the counter value for calculations
    workingCopy = creditCard;
    do
    {
        workingCopy /= 10;
        counter++;
    }
    while (workingCopy != 0);

    //Takes the length of the card number and returns a divisor to get the first digit
    for (int i = 0; i < counter - 2; i++)
    {
        divide *= 10;
    }

    //Uses the divisor to get the first or first two digits of the credit cards number for checks.
    int firstDigit = creditCard / divide;
    int firstTwo = creditCard / (divide / 10);

    //Checks if the credit card is valid and prints the corresponding credit card type.
    if (sum % 10 == 0)
    {
        if (firstDigit == 4 && (counter == 13 || counter == 16))
        {
            printf("VISA\n");
        }
        else if ((firstTwo == 34 || firstTwo == 37) && counter == 15)
        {
            printf("AMEX\n");
        }
        else if ((firstTwo >= 51 && firstTwo <= 55) && counter == 16)
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}
